driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2282,6 +2289,234 @@ public void run() {
+                    return new SchemaRefreshRequest(keyspace, null);
+                if(!this.table.equals(that.table))
+                    return new SchemaRefreshRequest(keyspace, null);
+                return this;
+            }
+
+            @Override
+            public String toString() {
+                if(this.keyspace == null)
+                    return "Refresh ALL";
+                if(this.table == null)
+                    return "Refresh keyspace " + keyspace;
+                return "Refresh keyspace " + keyspace + " table " + table;
+            }
+        }
+
+        private class SchemaRefreshRequestDeliveryCallback implements EventDebouncer.DeliveryCallback<SchemaRefreshRequest> {
+
+            @Override
+            public ListenableFuture<?> deliver(final List<SchemaRefreshRequest> events) {
+                return executor.submit(new ExceptionCatchingRunnable() {
+                    @Override
+                    public void runMayThrow() throws InterruptedException, ExecutionException {
+                        SchemaRefreshRequest coalesced = null;
+                        for (SchemaRefreshRequest request : events) {
+                            coalesced = coalesced == null ? request : coalesced.coalesce(request);
+                        }
+                        assert coalesced != null;
+                        logger.trace("Coalesced schema refresh request: {}", coalesced);
+                        controlConnection.refreshSchema(coalesced.keyspace, coalesced.table);
+                    }
+                });
+            }
+
+        }
+
+        private class NodeRefreshRequest {
+
+            private final Host host;
+
+            private final HostEvent eventType;
+
+            private NodeRefreshRequest(Host host, HostEvent eventType) {
+                this.host = host;
+                this.eventType = eventType;
+            }
+
+            @Override
+            public String toString() {
+                return host + " " + eventType;
+            }
+
+        }
+
+        private class NodeRefreshRequestDeliveryCallback implements EventDebouncer.DeliveryCallback<NodeRefreshRequest> {
+
+            @Override
+            public ListenableFuture<?> deliver(List<NodeRefreshRequest> events) {
+                Map<Host, HostEvent> hosts = new HashMap<Host, HostEvent>();
+                // only keep the last event for each host
+                for (NodeRefreshRequest req : events) {
+                    hosts.put(req.host, req.eventType);
+                }
+                List<ListenableFuture<?>> futures = new ArrayList<ListenableFuture<?>>(hosts.size());
+                for (final Entry<Host, HostEvent> entry : hosts.entrySet()) {
+                    final Host host = entry.getKey();
+                    HostEvent eventType = entry.getValue();
+                    switch (eventType) {
+                        case UP:
+                            schedule(hostUp(host));
```

Comment:
It looks like that is the case, however it seems that the only impact of this is that since `onAdd` is never called all the things that happen in there (lbp.onAdd, StateListener.onAdd, ControlConnection.onAdd).  I suppose the same would be the case for onRemove.

Here's some test code to produce this scenario, adding some log statements in NodeRefreshFrequestDeliveryCallback helps show what happens:

```java
    @Test(groups="short")
    public void should_call_onAdd_with_bootstrap_stop_start() {
        QueryOptions queryOptions = new QueryOptions().setRefreshNodeIntervalMillis(10000);
        CCMBridge ccm = CCMBridge.builder("test").withNodes(1).build();
        Cluster cluster = spy(Cluster.builder()
            .addContactPoint(CCMBridge.ipOfNode(1))
            .withQueryOptions(queryOptions)
            .build());

        try {
            cluster.connect();
            Host.StateListener listener = mock(Host.StateListener.class);
            cluster.register(listener);
            ccm.bootstrapNode(2);
            ccm.stop(2);
            ccm.start(2);
            // TODO: Ensure onAdd and onUp are called on listener.  Sleep for a while now to observe behavior in logs.
            Uninterruptibles.sleepUninterruptibly(100, TimeUnit.SECONDS);
        } finally {
            cluster.close();
            ccm.remove();
        }
    }
```

There are probably a number of ways to approach this, but the way I think this should work is that on NodeRefreshRequestDeliveryCallback:

1. retain at least 1 of each of [REMOVED, ADDED] if they exist.
2. retain the most recent (last) element in the list of events (whether it's up, down, removed, added)

This would ensure that the latest of up/down (if received) are processed and that any removed or added events are also processed.